src/importer.rs

```diff
@@ -89,65 +144,37 @@ impl Importer {
         let mut file = fs::File::open(path)?;
         let file_modified = file.metadata().and_then(|metadata| metadata.modified())?;
 
-        // If paths and timestamps match, we can just return cached models.
-        if let Some(path_metadata) = self.path_metadata.get(path) {
-            if path_metadata.last_modified == file_modified {
-                return Ok(self
-                    .loaded_models
-                    .get(&path_metadata.checksum)
-                    .expect("Should get loaded models by obj file's checksum")
-                    .clone());
-            }
-        }
-
-        let file_size = file.metadata().map(|m| m.len() as usize + 1).unwrap_or(0);
-        let mut file_contents = Vec::with_capacity(file_size);
-        file.read_to_end(&mut file_contents)?;
-        let checksum = calculate_checksum(&file_contents);
-
-        let models = match self.loaded_models.entry(checksum) {
-            Entry::Occupied(loaded_model) => {
-                self.path_metadata.insert(
+        let models = match self.cache.get_if_not_modified(path, file_modified) {
+            Some(models) => return Ok(models),
+            None => {
+                let file_size = file.metadata().map(|m| m.len() as usize + 1).unwrap_or(0);
+                let mut file_contents = Vec::with_capacity(file_size);
+                file.read_to_end(&mut file_contents)?;
+                let checksum = calculate_checksum(&file_contents);
+
+                let models = match self.cache.get_by_checksum(checksum) {
+                    Some(models) => models.clone(),
```

Comment:
Hmmm, shouldn't we have a way to just update the `last_modified` for a path in the cache in case the checksum checks out? Something like: we validated the cache entry is still valid for this timestamp. We can leave it for a FIXME, but this could prevent an allocation and de-allocation in the cache itself when you set new data.

Reply:
I'm not sure I understand what you mean. Matching checksum does only mean there are contents of some file cached, but it doesn't mean it has come from particular path. That's why path and metadata are cached anyway on line 163.

Reply:
On line 163 you set the models you acquired here back to the cache with new file_modified time. That is an additional drop on line 163 happening because you overwrite the data in cache with the same data. You could instead just say to the cache:

`cache.update_last_modified(path, file_modified)`, causing the cache to update it's metadata and not drop the data in cache only to be replaced with the same data.

Reply:
My bad, just realized the `set` method does different things in different contexts. Carry on!